### PR TITLE
Better checking for overload function/method declarations.

### PR DESCRIPTION
**Pull request by zhxchen17**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59956 [jit] Better checking for overload function/method declarations.**

Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

### CODE REVIEW

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

File: test/test_jit.py [hunk 1]
```diff
@@ -14168,6 +14168,26 @@ def test():
         with self.assertRaisesRegex(Exception, "Parameters not specified"):
             torch.jit.script(test)
 
+        def wrong_decl_body(x: str) -> str:
```
Comment (gmagogsfm) [inline test/test_jit.py:14171]: Nit:

Could you create a separate test case for it rather than extending existing test case? Long test case could make test failures less readable

File: torch/_jit_internal.py [hunk 2]
```diff
@@ -704,7 +703,17 @@ def copy_torchscript_modifier(orig, new) -> None:
...
 _overloaded_fns : Dict[str, List[Callable]] = {}  # noqa: T484
 
+def _check_overload_body(func):
+    parsed_def = parse_def(func)
+    body = parsed_def.ast.body[0].body
+    if len(body) != 1 or not isinstance(body[0], ast.Pass):
+        msg = "Only `pass` statement can be the body of overload declaration:\n"
+        msg += '\n'.join(parsed_def.source.split("\n")[:3])
+        msg += " <- Expecting `pass` here! "
+        raise RuntimeError(msg)
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:713]: When user hits this error, chances are they do not know how to use `_overload` correctly. Maybe it is worthwhile to go extra mile by showing the correct way of using `_overload` here.

File: torch/jit/_script.py [hunk 3]
```diff
@@ -1221,6 +1221,14 @@ def _get_overloads(obj):
     if uncompiled_overloads is None:
         return existing_compiled_fns
 
+    if obj in uncompiled_overloads:
+        sourcelines, file_lineno, filename = _jit_internal.get_source_lines_and_file(obj)
+        raise RuntimeError(
```
Comment (gmagogsfm) [inline torch/jit/_script.py:1226]: Same here, we might want to show complete usage guidance on `_overload` in this case.

File: test/test_jit.py [hunk 4]
```diff
@@ -14168,6 +14168,26 @@ def test():
...
+            pass
+
+        @torch.jit._overload
+        def null_overload(x: str) -> str:  # noqa: F811
+            pass
+
+        def null_overload_driver():
+            return null_overload(0)
+
+        with self.assertRaisesRegex(RuntimeError, 'Implementation for the function ".+" is missing.'):
```
Comment (gmagogsfm) [inline test/test_jit.py:14188]: What about the case, say there is a later definition of `null_overload` but its signature mismatches that of its overload declarations?

Comment (zhxchen17) [reply test/test_jit.py:14188]: we can add another check to test whether decl and def have the same # of arguments, I guess that's what you mean?

File: torch/jit/_script.py [hunk 5]
```diff
@@ -1221,6 +1221,14 @@ def _get_overloads(obj):
     if uncompiled_overloads is None:
         return existing_compiled_fns
 
+    if obj in uncompiled_overloads:
```
Comment (SplitInfinity) [inline torch/jit/_script.py:1224]: I am not super familiar with this code. Why does this check imply that there is no definition for the function?

File: torch/jit/_script.py [hunk 6]
```diff
@@ -1221,6 +1221,14 @@ def _get_overloads(obj):
...
+        sourcelines, file_lineno, filename = _jit_internal.get_source_lines_and_file(obj)
+        raise RuntimeError(
+            f'Implementation for the function "{qual_name}" is missing. Please make sure a ' +
+            "definition is provided and defined after all overload declarations.\n" +
+            f'File "{filename}", line {file_lineno}:' + "\n" + ''.join(sourcelines)
```
Comment (SplitInfinity) [inline torch/jit/_script.py:1229]: Is it possible to use `SourceRange` here to highlight source in the same way that we do for exceptions through during compilation of the JIT AST?

File: test/test_jit.py (hunk 1 above)
Comment (SplitInfinity) [reply test/test_jit.py:14171]: Another option is to use [subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests).

**Review by SplitInfinity — COMMENTED** [review]
Nice!

File: torch/jit/_script.py (hunk 6 above)
Comment (zhxchen17) [reply torch/jit/_script.py:1229]: These checks are performed before we "compile" python asts, so the source ranges are unavailable at this point. Potentially we can move this check into the actual compilation steps, but I find that doesn't make the program simpler, so I'm inclined to a bit straightforward error checking logic rather than a little better error message, assuming this error is likely to be rare.

File: torch/_sources.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+@functools.lru_cache(maxsize=None)
+def make_source_context(*args):
+    return SourceContext(*args)
+
+
+def fake_range():
+    return SourceContext('', None, 0, 0).make_raw_range(0, 1)
+
+
+ParsedDef = namedtuple('ParsedDef', ['ast', 'ctx', 'source'])
```
[the remark this answers is not in the document]
Comment (zhxchen17) [reply torch/_sources.py:88]: didn't know this is a thing! yeah NamedTuple is a lot better.

File: torch/jit/_script.py (hunk 5 above)
Comment (zhxchen17) [reply torch/jit/_script.py:1224]: If user put "@overload def foo():" as the last toplevel item for overload `foo` i.e. no definition, then jit compiler will recognize that as the "definition" since it's the last defined item in python with that name, so that decl function will appear as `obj` here.

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

File: torch/_jit_internal.py [hunk 8]
```diff
@@ -716,7 +715,50 @@ def copy_torchscript_modifier(orig, new) -> None:
...
+    if isinstance(x, type0):
+        return x
+    elif isinstance(x, type1):
+        return x
+'''
+
+def get_overload_no_implementation_error_message(kind, obj):
+    sourcelines, file_lineno, filename = get_source_lines_and_file(obj)
+    return (
+        f'Implementation for the {kind} "{_qualified_name(obj)}" is missing. Please make'
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:739]:
```suggestion
        f'Implementation for the {kind} "{_qualified_name(obj)}" is missing. Please make '
```

File: torch/_jit_internal.py [hunk 9]
```diff
@@ -716,7 +715,50 @@ def copy_torchscript_modifier(orig, new) -> None:
...
+        f'sure a definition is provided and defined after all overload declarations.\n'
+        f'File "{filename}", line {file_lineno}:\n' + ''.join(sourcelines) + _OVERLOAD_EXAMPLE
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:741]: Nit:
I'd add a few empty lines after printing `sourcelines` to prevent `_OVERLOAD_EXAMPLE` from being mixed in user source code.

File: torch/_jit_internal.py [hunk 10]
```diff
@@ -716,7 +715,50 @@ def copy_torchscript_modifier(orig, new) -> None:
...
+
+    def is_pass(x):
+        return isinstance(x, ast.Pass)
+
+    def is_ellipsis(x):
+        return isinstance(x, ast.Expr) and isinstance(x.value, ast.Ellipsis)
+
+    if len(body) != 1 or not (is_pass(body[0]) or is_ellipsis(body[0])):
+        msg = "Only `pass` statement or `...` can be the body of overload declaration:\n"
+        msg += '\n'.join(parsed_def.source.split("\n")[:3])
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:756]: Looks like this is pointing to a segment of problematic code, we have some utility for it:

https://github.com/pytorch/pytorch/blob/35307b131df9d24bfa96103d6061cf14c797ee32/torch/csrc/jit/frontend/source_range.h#L122

Is it possible to use the common `SourceRange` here as well?

File: torch/_jit_internal.py [hunk 11]
```diff
@@ -994,22 +1037,6 @@ def _qualified_name(obj) -> str:
     return module_name + "." + name
 
 
-# Thin wrapper around SourceRangeFactory to store extra metadata
```
Comment (gmagogsfm) [inline torch/_jit_internal.py:997]: Is this code block just moved or moved+modified? 

I think this refactor (along with others you made in this PR) makes sense, just that in general, it is slightly better to put refactoring in a separate PR whenever possible so that each PR has a clear purpose.

**Review by gmagogsfm — APPROVED** [review]
Looks good, just have a few nits.

File: torch/_jit_internal.py (hunk 11 above)
Comment (zhxchen17) [reply torch/_jit_internal.py:997]: it's just moved.

**Comment by zhxchen17** [comment]
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).
